File: 2021/julia/Day4.jl
module Day4

export Submarine, on_input!, output

@enum ReadState ReadingChosenNumbers ReadingBoards

mutable struct BingoBoard
    nums::Vector{Int32}
    won::Bool
    win_score::Int32

    BingoBoard() = new([], false, 0)
end

mutable struct Submarine
    input_read_state::ReadState
    current_read_board::Union{BingoBoard, Nothing}

    chosen_bingo_numbers::Vector{Int32}
    bingo_boards::Vector{BingoBoard}

    Submarine() = new(ReadingChosenNumbers, nothing, [], [])
end

function on_input!(s::Submarine, a::AbstractString)
    on_input!(s::Submarine, Val(s.input_read_state), a)
end

function on_input!(s::Submarine, ::Val{ReadingChosenNumbers}, a::AbstractString)
    if a == ""
        s.input_read_state = ReadingBoards
        s.current_read_board = BingoBoard()
    else
        s.chosen_bingo_numbers = split(a, ",") .|> (x) -> parse(Int32, x)
    end
end

function on_input!(s::Submarine, ::Val{ReadingBoards}, a::AbstractString)
    if a == ""
        push!(s.bingo_boards, something(s.current_read_board))
        s.current_read_board = BingoBoard()
    else
        nums = filter(!isempty, split(a, " ")) .|> (x) -> parse(Int32, x)
        s.current_read_board.nums = vcat(s.current_read_board.nums, nums)
    end
end

function mark_number(val::Int32, num::Int32)
    val == num ? -val : val
end

function mark_board!(b::BingoBoard, num::Int32)
    b.nums = b.nums .|> (x) -> mark_number(x, num)
end

function has_board_won(b::BingoBoard)::Bool
    for i in 1:5
        row_win = all(signbit, Iterators.take(Iterators.drop(b.nums, (i - 1) * 5), 5))
        column_win = all(signbit, Iterators.flatten(Iterators.partition(b.nums, 5) .|> (r) -> first(Iterators.drop(r, i - 1))))

        if row_win || column_win
            return true
        end
    end

    false
end

function sum_unmarked(b::BingoBoard)::Int32
    sum(+, filter(!signbit, b.nums))
end

# This should've been named output! because for this particular situation, running through the code will mutate `s`. If I did this, I'd have to go back and rename the previous days' `output` functions as well, but that would've meant that they're now incorrectly named, because they don't mutate `s`. I'm too lazy to figure out a solution for this right now, so `output` it is.
function output(s::Submarine)
    first_board_score = 0
    last_board_score = 0

    for num in s.chosen_bingo_numbers
        foreach(mark_board!, s.bingo_boards, Iterators.repeated(num))

        won_boards = filter(has_board_won, s.bingo_boards)

        if first_board_score == 0 && length(won_boards) > 0
            first_board_score = num * sum_unmarked(won_boards[1])
        end

        if length(won_boards) == 1 && length(s.bingo_boards) == 1
            last_board_score = num * sum_unmarked(won_boards[1])
            break
        end

        s.bingo_boards = filter(!has_board_won, s.bingo_boards)
    end

    println("Part 1: $(first_board_score)")
    println("Part 2: $(last_board_score)")
end

end

File: 2021/rust/src/day4.rs
struct BingoBoard {
    // Negative number means it's marked.
    pub numbers: Vec<i32>,
    won: bool,
    win_score: i32,
}

impl BingoBoard {
    pub fn new() -> Self {
        Self {
            numbers: Vec::new(),
            won: false,
            win_score: 0,
        }
    }

    pub fn mark(&mut self, number: i32) {
        self.numbers.iter_mut().for_each(|n| {
            if *n == number {
                *n = -number;
            }
        });

        self.evaluate_win(number);
    }

    pub fn evaluate_win(&mut self, number: i32) {
        for i in 0..5 {
            let row_win = self.numbers.iter().skip(i * 5).take(5).all(|&x| x < 0);
            let column_win = self.numbers.iter().skip(i).step_by(5).all(|&x| x < 0);

            if row_win || column_win {
                self.won = true;
                self.win_score = number * self.numbers.iter().filter(|&&x| x > 0).sum::<i32>();
                // println!("Board won! Win score: {}", self.win_score);
            }
        }
    }
}

pub struct Submarine {
    input_read_state: InputReadState,
    current_read_board: Option<BingoBoard>,

    chosen_bingo_numers: Vec<i32>,
    bingo_boards: Vec<BingoBoard>,
}

enum InputReadState {
    ReadingChosenNumbers,
    ReadingBoards,
}

impl Submarine {
    pub fn new() -> Self {
        Self {
            input_read_state: InputReadState::ReadingChosenNumbers,
            current_read_board: None,

            chosen_bingo_numers: Vec::new(),
            bingo_boards: Vec::new(),
        }
    }

    pub fn on_input(&mut self, result : &str) {
        match self.input_read_state {
            InputReadState::ReadingChosenNumbers => {
                if result.is_empty() {
                    self.input_read_state = InputReadState::ReadingBoards;
                    self.current_read_board = Some(BingoBoard::new());
                } else {
                    self.chosen_bingo_numers = result.split(",").map(|x| x.parse().unwrap()).collect();
                }
            }
            InputReadState::ReadingBoards => {
                if result.is_empty() {
                    self.bingo_boards.push(self.current_read_board.take().unwrap());
                    self.current_read_board = Some(BingoBoard::new());
                } else {
                    self.current_read_board.as_mut().unwrap().numbers.extend(result.split(" ").filter(|x| !x.is_empty()).map(|x| x.parse::<i32>().unwrap()));
                }
            }
        }
    }

    pub fn output(mut self) {
        let mut last_board_score = 0;
        let mut first_board_score = 0;

        for num in self.chosen_bingo_numers.iter() {
            self.bingo_boards.iter_mut().for_each(|b| b.mark(*num));

            match self.bingo_boards.iter().find(|b| b.won) {
                Some(board) => {
                    if first_board_score == 0 {
                        first_board_score = board.win_score;
                    }

                    if self.bingo_boards.len() == 1 {
                        last_board_score = board.win_score;
                        break;
                    }
                },
                None => {}
            };

            self.bingo_boards = self.bingo_boards.into_iter().filter(|b| !b.won).collect();
        }

        println!("Part 1: {}", first_board_score);
        println!("Part 2: {}", last_board_score);
    }
}